### PR TITLE
coprocessor/*: support streaming

### PR DESCRIPTION
**Pull request by AndreMouche**
Hi,
      This PR add an new interface for coprocessor to support streaming, the related PR in kvproto is https://github.com/pingcap/kvproto/pull/202

@hicqu @BusyJay PTAL

### CODE REVIEW

File: src/coprocessor/dag/dag.rs [hunk 1]
```diff
@@ -58,38 +59,64 @@ impl<'s> DAGContext<'s> {
...
+        let is_stream = on_resp.is_stream();
         loop {
+            if is_stream && cur_chunk_count >= BATCH_ROW_COUNT {
+                let mut resp = Response::new();
+                let mut sel_resp = SelectResponse::new();
+                sel_resp.set_chunks(RepeatedField::from_vec(chunks));
+                let data = box_try!(sel_resp.write_to_bytes());
+                resp.set_data(data);
+                on_resp.resp(resp);
+                chunks = vec![];
```
Comment (siddontang) [inline src/coprocessor/dag/dag.rs]: `chunks.clear()`

File: src/server/service/kv.rs [hunk 2]
```diff
@@ -55,6 +60,28 @@ pub struct Service<T: RaftStoreRouter + 'static> {
...
+        StreamChunk { rc: rc }
+    }
+}
+
+impl<T: Debug + Send + 'static> Stream for StreamChunk<T> {
+    type Item = (T, WriteFlags);
+    type Error = grpc::Error;
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        match self.rc.recv().unwrap() {
+            Some(resp) => Ok(Async::Ready(Some((resp, WriteFlags::default())))),
```
Comment (siddontang) [inline src/server/service/kv.rs]: seem you should use BufferHint here, then flush sometimes.

File: src/server/service/kv.rs [hunk 3]
```diff
@@ -767,6 +815,41 @@ impl<T: RaftStoreRouter + 'static> tikvpb_grpc::Tikv for Service<T> {
...
+        ctx: RpcContext,
+        req: Request,
+        sink: ServerStreamingSink<Response>,
+    ) {
+        let label = "coprocessor_stream";
+        let timer = GRPC_MSG_HISTOGRAM_VEC
+            .with_label_values(&[label])
+            .start_coarse_timer();
+
+        let (cb, future) = make_stream_callback();
```
Comment (siddontang) [inline src/server/service/kv.rs]: `(cb, stream)` is better here.

File: src/coprocessor/dag/dag.rs [hunk 4]
```diff
@@ -58,38 +59,64 @@ impl<'s> DAGContext<'s> {
...
+        on_resp: &'s mut OnResponse,
+    ) -> Result<Option<Response>> {
         self.validate_dag()?;
         let mut exec = self.build_dag(statistics)?;
         let mut chunks = vec![];
+        let mut cur_chunk_count = 0;
+        let is_stream = on_resp.is_stream();
         loop {
+            if is_stream && cur_chunk_count >= BATCH_ROW_COUNT {
+                let mut resp = Response::new();
```
Comment (ngaut) [inline src/coprocessor/dag/dag.rs]: DRY. See line 106

File: src/coprocessor/dag/dag.rs [hunk 5]
```diff
@@ -220,3 +241,13 @@ fn inflate_cols(row: &Row, cols: &[ColumnInfo], output_offsets: &[u32]) -> Resul
     }
     Ok(values)
 }
+
+fn response_from_chunks(chunks: Vec<Chunk>) -> Result<Response> {
+
```
Comment (siddontang) [inline src/coprocessor/dag/dag.rs]: remove the blank line.

File: src/coprocessor/dag/dag.rs [hunk 6]
```diff
@@ -58,38 +60,59 @@ impl<'s> DAGContext<'s> {
...
         let mut exec = self.build_dag(statistics)?;
         let mut chunks = vec![];
+        let mut cur_chunk_count = 0;
+        let is_stream = on_resp.is_stream();
         loop {
+            if is_stream && cur_chunk_count >= BATCH_ROW_COUNT &&
+                chunks.len() >= CHUNKS_PER_STREAM
+            {
+                let mut cur_chunks: Vec<Chunk> = vec![];
+                chunks = mem::replace(&mut cur_chunks, chunks);
```
Comment (siddontang) [inline src/coprocessor/dag/dag.rs]: any benefit to using replace here?

**Pull request closed without merging**